### DISCUSSION TITLE
Clangd roadmap?

### DISCUSSION THREAD

**Discussion by HighCommander4** [clangd/clangd#766] [Ideas]
Clangd gets contributions from a variety of sources. Some are from volunteers and some from people paid to work on clangd by various organizations.

I think it would be valuable for the community (both contributors and users) to gain some insight into what regular (or even occasional) contributors **plan to work on** in the upcoming [whatever your planning time-horizon is, likely to vary from person to person]. (Obviously, these would be non-binding outlines of plans, we all know priorities and circumstances change over time.)

In addition, I think it would be valuable to share what are some areas of improvement people think the project **needs**, even if they're not things you plan to or are able to work on yourself. These could be accumulated into a list (less granular than "all open issues") that could serve as inspiration for contributors _looking for_ something to work on.

Over time, these lists of plans and needs could be massaged into a project **roadmap** of some sort.

**Comment by HighCommander4** [comment]
I'm happy to start by sharing my own plans and opinions about project needs.

**Plans**

Here are some of the things I hope to tackle in the coming months:

 * Continue work on inlay hints. In addition to the in-progress [parameter hints](https://github.com/clangd/clangd/issues/313), I'd like to explore type hints and implicit conversion hints.
 * Further improvements to editor features working (heuristically) in templated code, such as:
   * https://github.com/clangd/clangd/issues/432
   * https://github.com/clangd/clangd/issues/451
   * https://github.com/clangd/clangd/issues/443
   * Others which haven't been filed yet. Ideally, I'd like to achieve parity with Eclipse CDT here, though, since we're talking about heuristics, there will always be room for further improvements.
 * A couple of improvements to find-references which depend on a [protocol enhancement](https://github.com/microsoft/language-server-protocol/issues/396):
   * Annotating references with the name of the function in which they occur
   * Annotating references as read- or write-, so that clients can filter and show e.g. only write-references.
 * Perhaps some improvements to responsiveness on file-open, such as https://github.com/clangd/clangd/issues/506

**Needs**

 * I think clangd could use further improvements to responsiveness on file-open.
   * There have been discussions of pseudo-parser based approaches which sound promising, but this is a project whose scale is a bit beyond me to attempt personally.
   * Alternatively, C++20 Modules might bring a lot of the desired improvements (since, instead of parsing headers, clangd will only need to deserialize compiled module interface files). The downside is, the improvement will only be realized to the extent that a codebase has actually transitioned to using C++20 Modules, and that adoption might take a while, especially for large projects.

**Comment by kadircet** [comment]
Thanks! This was something we've discussed internally a lot but never get to do. Mostly because of maintenance burden and our planning being split between open-source clangd-related stuff and improvements to $Employer's other c++ tooling needs.

Main questions that pop into my mind are:
- Where do we keep this roadmap?
- Who can update it?
- How granular should it be?
- Should it contain "future" ideas or limited to stuff people are "actively" working on.

I feel like it might be more feasible to have a markdown file checked in to either clangd/clangd or llvm-project/llvm talking about what people are "actively" working on. Rather than a roadmap, we can have a "Stuff we are working on" section that any contributors can send patches to update and maintainers will approve as usual (the approval here means maintainers are also committing to review changes, participate in design discussions and maintain the result afterwards).
On the granularity side, I think any systematic improvement (not just one-time bug fixes, but a series of ongoing changes that'll probably take more than a couple weeks to finalize) to clangd can go into this bucket.

As for why:
- Keeping the roadmap here as a discussion(or an issue) probably means stuff won't be visible after a certain period(due to noise) and people won't be able to find things easily.
- Dedicating to a roadmap rather than a "stuff we are actively working on" means items might get stale/stuck. Also promising for something much bigger that won't happen immediately requires a lot more effort and coordination that I don't think maintainers can afford.
- Trying to update the document for any improvement that'll be landed in less than a week or so wouldn't really be worth anyone's time. I think such changes should still be carried on with an issue/discussion in clangd/clangd repo.

As for "needs":
- I agree that these should have better definition than "any open issue".
- Who can put entries into this list?
- What do maintainers promise about fulfilling this backlog?

This is much riskier and harder to maintain. As the items in this list would need to be prioritized somehow. For example it wouldn't be really desired by maintainers to work on (either drive the change or just provide review/design input) some project that'll impact only a small set of users when there are other items that can benefit a much larger user base.
This is unfortunately a hard-to-draw line. Any solution I can think of requires lots of discussions both internally and externally which will imply lots of maintenance burden for very small value in return.

Maybe we can have a list of maintainer approved "needs" (that require multi-week commitment) which can guide new contributors, saying this is by no means a binding list that people should feel forced to, but more of a "patches welcome" list.

**Reply by HighCommander4** [reply]
Thanks for your thoughts!

I agree that a "stuff we are working on" list would be a great start.

As an alternative to a checked-in document, we could consider making it a Wiki page. I think that's somewhat common (e.g. that's where vscode keeps [their roadmap](https://github.com/microsoft/vscode/wiki/Roadmap)), and probably reduces the friction to keeping it updated.

I do think there's value in the added information that "maintainers are also committing to review changes, participate in design discussions and maintain the result afterwards", but perhaps that can be communicated via an indication on the wiki page, rather than through code review? For example, the page could be a table with three columns: "Feature / Item", "Who is working on it", and "Maintainer interest", with the last one being a checkbox that maintainers can fill in?

---

In terms of the "needs" section, I think there are two main use cases for it:

  * As ideas for people who'd like to contribute but not sure what to work on. For this use case, I think it's important that the list be:
     * Less granular than issues, as discussed
     * Prioritized
     * Not include things that are on the "stuff we are working on" list (this is why I sort of envisioned them as being two sections of the same page / document)
  * To the extent that some $Employers wish to invest in making clangd better in general, having a list of things that the community agrees are high-priority helps make the case for working on those things.

To keep the maintenance burden low, I think this list can be fairly high-level, and initially be populated / edited by maintainers only, based on pain points that have come up repeatedly in issues and things like that.

**Comment by HighCommander4** [comment]
As an example, I think "Google is investing time in 2021 into the development of a [protoype for a pseudo-parser](https://lists.llvm.org/pipermail/cfe-dev/2021-November/069246.html) with the aim of making some clangd features more responsive" would have been a worthwhile piece of information to communicate to the broader community, using something like a roadmap. (Exciting project btw!)

**Reply by HighCommander4** [reply]
Sorry to harp on this, but reading through the [pseudo-parser discussion thread](https://discourse.llvm.org/t/rfc-a-c-pseudo-parser-for-tooling/59217/42) (which is super interesting and I'm learning a lot from!), it occurs to me that insights like this:

> You could construct a syntactic model (clang::syntax::Tree) from a clang AST, and this was the first thing we tried. (buildSyntaxTree). This made a nice demo but we couldn’t make it robust without building most of a parser and using the AST as a guide, at which point why make the AST mandatory?

are very interesting and valuable high-level design insights, and it would really benefit the community to have them communicated in a public venue (more visible than a Phabricator comment) at the time they occur.
